### PR TITLE
fix unstable unit test and polish ReadOnlyTsFile query logic

### PR DESCRIPTION
**Pull request by LeiRui**

### CODE REVIEW

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/TimeRange.java [hunk 1]
```diff
@@ -183,7 +184,19 @@ public static boolean overlaps(long minA, long maxA, long minB, long maxB) {
 
   @Override
   public String toString() {
-    return "[ " + min + " : " + max + " ]";
+    String res;
```
Comment (jixuan1989) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/TimeRange.java]: StringBuilder

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/TimeRange.java [hunk 2]
```diff
@@ -207,6 +220,38 @@ public boolean getRightClose() {
...
+  public static List<TimeRange> getUnions(ArrayList<TimeRange> unionCandidates) {
+    ArrayList<TimeRange> unionResult = new ArrayList<>();
+    Iterator<TimeRange> iterator = unionCandidates.iterator();
+    TimeRange range_curr;
+
+    if (!iterator.hasNext()) {
+      return unionResult;
+    } else {
+      TimeRange r = iterator.next();
+      range_curr = new TimeRange(r.getMin(), r.getMax());
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/TimeRange.java]:
```suggestion
      range_curr = iterator.next();
```

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/TimeRange.java [hunk 3]
```diff
@@ -207,6 +220,38 @@ public boolean getRightClose() {
...
+    }
+
+    while (iterator.hasNext()) {
+      TimeRange range_next = iterator.next();
+      if (range_curr.intersects(range_next)) {
+        range_curr.set(Math.min(range_curr.getMin(), range_next.getMin()),
+            Math.max(range_curr.getMax(), range_next.getMax()));
+      } else {
+        unionResult.add(new TimeRange(range_curr.getMin(), range_curr.getMax()));
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/TimeRange.java]:
```suggestion
        unionResult.add(range_curr);
```

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/TimeRange.java [hunk 4]
```diff
@@ -207,6 +220,38 @@ public boolean getRightClose() {
...
+        range_curr.set(range_next.getMin(), range_next.getMax());
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/TimeRange.java]:
```suggestion
        range_curr = range_next;
```

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/TimeRange.java [hunk 5]
```diff
@@ -207,6 +220,38 @@ public boolean getRightClose() {
...
+      }
+    }
+    unionResult.add(new TimeRange(range_curr.getMin(), range_curr.getMax()));
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/TimeRange.java]:
```suggestion
    unionResult.add(range_curr);
```

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/controller/MetadataQuerier.java [hunk 6]
```diff
@@ -54,35 +54,27 @@
...
-   * @throws IOException
+   * @param paths given paths
+   * @param targetMode either InPartition or PrevPartition
+   * @param partitionStartOffset the start offset of the partition
+   * @param partitionEndOffset the end offset of the partition
+   * @return the union time ranges in ascending order.
    */
-  public ArrayList<TimeRange> getTimeRangeInOrPrev(List<Path> paths, LoadMode targetMode)
-      throws IOException;
+  ArrayList<TimeRange> getTimeRangeInOrPrev(List<Path> paths, LoadMode targetMode,
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/controller/MetadataQuerier.java]: add @Override annotation

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/controller/MetadataQuerierByFileImpl.java [hunk 7]
```diff
@@ -52,50 +50,18 @@
 
   private TsFileSequenceReader tsFileReader;
 
-  private LoadMode mode;
+  private LoadMode mode; // be the default mode (NoPartition) most of the time
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/controller/MetadataQuerierByFileImpl.java]:
```suggestion
  // be the default mode (NoPartition) most of the time
  private LoadMode mode;
```

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/controller/MetadataQuerierByFileImpl.java [hunk 8]
```diff
@@ -52,50 +50,18 @@
...
-
-    if (!params.containsKey(QueryConstant.PARTITION_START_OFFSET) || !params
-        .containsKey(QueryConstant.PARTITION_END_OFFSET)) {
-      throw new IllegalArgumentException(
-          "Input parameters miss partition_start_offset or partition_end_offset");
-    }
-    this.mode = LoadMode.InPartition;
-    this.partitionStartOffset = params.get(QueryConstant.PARTITION_START_OFFSET);
-    this.partitionEndOffset = params.get(QueryConstant.PARTITION_END_OFFSET);
+    this.mode = LoadMode.NoPartition; // default mode
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/controller/MetadataQuerierByFileImpl.java]:
```suggestion
    // default mode
    this.mode = LoadMode.NoPartition;
```

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/controller/MetadataQuerierByFileImpl.java [hunk 9]
```diff
@@ -263,56 +233,37 @@ public TSDataType getDataType(String measurement) throws NoMeasurementException
...
-            Math.max(range_curr.getMax(), range_next.getMax()));
-      } else {
-        unionResult.add(new TimeRange(range_curr.getMin(), range_curr.getMax()));
-        range_curr.set(range_next.getMin(), range_next.getMax());
-      }
-    }
-    unionResult.add(new TimeRange(range_curr.getMin(), range_curr.getMax()));
-
-    this.mode = LoadMode.InPartition; // restore the mode to InPartition
+    this.mode = LoadMode.NoPartition; // restore the default mode. DO NOT REMOVE THIS LINE.
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/controller/MetadataQuerierByFileImpl.java]:
```suggestion
    // restore the default mode. DO NOT REMOVE THIS LINE.
    this.mode = LoadMode.NoPartition;
```

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/controller/MetadataQuerierByFileImpl.java [hunk 10]
```diff
@@ -263,56 +233,37 @@ public TSDataType getDataType(String measurement) throws NoMeasurementException
...
+   * partition constraints. Used to control the behavior of the loadChunkMetaData function.
+   *
    * @param chunkGroupMetaData a chunk group's metadata
-   * @return True if the chunk group can be accessed. False otherwise.
+   * @return True if the chunk group can be accessed under the current load mode. False otherwise.
    * @throws IOException illegal mode
    */
   private boolean checkAccess(ChunkGroupMetaData chunkGroupMetaData) throws IOException {
+    if (mode == LoadMode.NoPartition) {
+      return true; // always true
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/controller/MetadataQuerierByFileImpl.java]:
```suggestion
      // always true
      return true;
```

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/query/executor/TsFileExecutor.java [hunk 11]
```diff
@@ -136,6 +83,71 @@ public QueryDataSet execute(QueryExpression queryExpression) throws IOException
...
   }
 
+  /**
+   * Query with the partition constraint.
+   *
+   * @param queryExpression query expression
+   * @param params [partition_start_offset:long, partition_end_offset:long]
+   * @return QueryDataSet
+   */
+  public QueryDataSet execute(QueryExpression queryExpression, HashMap<String, Long> params) throws IOException {
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/query/executor/TsFileExecutor.java]:
```suggestion
  public QueryDataSet execute(QueryExpression queryExpression, Map<String, Long> params) throws IOException {
```

File: tsfile/src/test/java/org/apache/iotdb/tsfile/read/ReadInPartitionTest.java [hunk 12]
```diff
@@ -88,107 +130,129 @@ public void after() throws IOException {
...
-    QueryDataSet queryDataSet = roTsFile.query(queryExpression);
+    QueryDataSet queryDataSet = roTsFile.query(queryExpression, params);
+    IExpression transformedExpression = queryExpression
+        .getExpression(); // the expression is transformed
+
+    // test the transformed expression
+    Assert.assertEquals(ExpressionType.GLOBAL_TIME, transformedExpression.getType());
+
+    IExpression oracleTimeExpression = d1s6timeRangeList.get(0).getExpression();
+    String oracle = ExpressionOptimizer.getInstance().optimize(oracleTimeExpression,
```
Comment (qiaojialin) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/read/ReadInPartitionTest.java]: what is 'oracle'? try to use a straightforward name

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/TimeRange.java (hunk 3 above)
Comment (LeiRui) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/TimeRange.java]: No. Have you noticed that the next line changes the contents of the range_curr?

Comment (LeiRui) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/TimeRange.java]: Ok. I got it.

**Pull request merged**